snitun/server/worker.py

```diff
@@ -0,0 +1,119 @@
+from threading import Thread
+from typing import Dict, Optional, List, Tuple
+from socket import socket
+
+from .listener_peer import PeerListener
+from .listener_sni import SNIProxy
+from .peer_manager import PeerManager, PeerManagerEvent
+from .peer import Peer
+
+
+class ServerWorker(Process):
+    """Worker for multiplexer."""
+
+    def __init__(
+        self,
+        fernet_keys: List[str],
+        throttling: Optional[int] = None,
+    ) -> None:
+        """Initialize worker & communication."""
+        super().__init__()
+
+        self._fernet_keys: List[str] = fernet_keys
+        self._throttling: Optional[int] = throttling
+
+        # Used on the child
+        self._peers: Optional[PeerManager] = None
+        self._list_sni: Optional[SNIProxy] = None
+        self._list_peer: Optional[PeerListener] = None
+        self._loop: Optional[asyncio.BaseEventLoop] = None
+
+        # Communication between Parent/Child
+        self._manager: Manager = Manager()
+        self._new: Queue = self._manager.Queue()
+        self._sync: Dict[str, None] = self._manager.dict()
+
+    @property
+    def peer_size(self) -> int:
+        """Return amount of managed peers."""
+        return len(self._sync)
+
+    def is_responsible_peer(self, sni: str) -> bool:
+        """Return True if worker is responsible for this peer domain."""
+        return sni in self._sync
+
+    async def _async_init(self) -> None:
+        """Initialize child process data."""
+        self._peers = PeerManager(
+            self._fernet_keys,
+            throttling=self._throttling,
+            event_callback=self._event_stream,
+        )
+        self._list_sni = SNIProxy(self._peers)
+        self._list_peer = PeerListener(self._peers)
+
+    def _event_stream(self, peer: Peer, event: PeerManagerEvent) -> None:
+        """Event stream peer connection data."""
+        if event == PeerManagerEvent.CONNECTED:
+            self._sync[peer.hostname] = None
+        else:
+            self._sync.pop(peer.hostname, None)
+
+    def shutdown(self) -> None:
+        """Shutdown child process."""
+        self._new.put(None)
+        self.join(10)
+
+    def handover_connection(
+        self, con: socket, data: bytes, sni: Optional[str] = None
+    ) -> None:
+        """Move new connection to worker."""
+        self._new.put_nowait((con, data, sni))
+
+    def run(self) -> None:
+        """Running worker process."""
+        self._loop = asyncio.new_event_loop()
+
+        # Start eventloop
+        running_loop = Thread(target=self._loop.run_forever)
+        running_loop.start()
+
+        # Init backend
+        asyncio.run_coroutine_threadsafe(self._async_init(), loop=self._loop).result()
+
+        while True:
+            new: Tuple[socket, bytes, Optional[str]] = self._new.get()
+            if new is None:
+                break
+
+            new[0].setblocking(False)
+            asyncio.run_coroutine_threadsafe(
```

Comment:
This creates a future for the result, but you're not using that.

This is what `run_coroutine_threadsafe` does to schedule the coroutine:

```suggestion
            self._loop.call_soon_threadsafe(
                asyncio.ensure_future,
```

You might want to store `self._loop` as `loop` in the local context to make it faster.

Reply:
I looked into docs: https://docs.python.org/3/library/asyncio-task.html#asyncio.create_task is the new function

Reply:
they have no loop argument, so we not exactly know what will be happening because we also not know what comes from the parent/fork process.